cloud-controller-manager/do/doks_lb_validator.go

```diff
@@ -0,0 +1,206 @@
+type DOKSLBServiceValidator struct {
+	decoder *admission.Decoder
+	Log     logr.Logger
+	gClient *godo.Client
+}
+
+func initDOClient() (doClient *godo.Client, err error) {
+	token := os.Getenv(doAccessTokenEnv)
+
+	opts := []godo.ClientOpt{}
+
+	opts = append(opts, godo.SetUserAgent("digitalocean-webhook-server/"+version))
+
+	if token == "" {
+		return nil, fmt.Errorf("environment variable %q is required", doAccessTokenEnv)
+	}
+
+	tokenSource := &tokenSource{
+		AccessToken: token,
+	}
+
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	doClient, err = godo.New(oauthClient, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create godo client: %s", err)
+	}
+
+	return doClient, nil
+}
+
+// DOKSLBServiceValidator ...
+func (v *DOKSLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+	v.Log.V(10).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "bad request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	v.Log.V(10).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer {
+		return admission.Allowed("")
+	}
+
+	// check if request is for LB deletion
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer && svc.DeletionTimestamp != nil {
+		return admission.Allowed("the loadbalancer is deleting")
+	}
+
+	// initialize DO Client if not already initialized
+	if v.gClient == nil {
+		doClient, err := initDOClient()
+		if err != nil {
+			v.Log.Error(err, "failed to initialize DO client")
+			return admission.Errored(http.StatusConflict, err)
+		}
+		v.gClient = doClient
+	}
+
+	clusterID := os.Getenv(doClusterIDEnv)
+	if clusterID == "" {
+		fmt.Println("missing cluster id")
+	}
+	region := "nyc1"
+	if region == "" {
+		fmt.Println("missing region id")
+	}
+
+	nodePools, _, err := v.gClient.Kubernetes.ListNodePools(ctx, clusterID, &godo.ListOptions{})
+	if err != nil {
+		v.Log.Error(err, "no nodes found")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	var dropletIDs []int
+	for _, pool := range nodePools {
+		for _, node := range pool.Nodes {
+			dropletID, err := strconv.Atoi(node.DropletID)
+			if err != nil {
+				v.Log.Error(err, "failed to retrieve droplet ids")
+				return admission.Errored(http.StatusConflict, err)
+			}
+			dropletIDs = append(dropletIDs, dropletID)
+		}
+	}
+
+	forwardingRules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	lbRequest, err := v.buildRequest(svc.Name, region, dropletIDs, forwardingRules)
+	if err != nil {
+		v.Log.Error(err, "failed to build lb request")
+		//fmt.Printf("failed to build lb request", err)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	err = v.decoder.DecodeRaw(req.OldObject, svc)
+	if err != nil {
```

Comment:
Is there a way to distinguish between genuine recoding errors and the old object not existing? For instance, is `req.OldObject` possibly nil when the Service is brand new?

Besides, we cannot assume that absence of the old object alone implies that a create request is needed: for instance, users trying to create an LB with an erroneous configuration predating the webhook would have a Service object but the LB might have never been created. For such a case, we would want to send out a create validation request.

 I think it's more reliable to look at the status field of the old Service object (if it exists) and determine if we have a field set indicating that an LB create has started. An easy way to discover these fields is to create a managed LB and check the backing Service object right after that.

Reply:
@time req.OldObject will error if empty:

> It errors out if rawObj is empty i.e. containing 0 raw bytes.

From an existing LB, the only field under the status is the loadbalancer IP:
```
status:
  loadBalancer:
    ingress:
    - ip: 144.126.249.45
```
There is a creationTimestamp in the metadata, perhaps we can use that instead?

Reply:
The IP would only be available once an LB is fully provisioned, so I'm not sure we can use that.

Can you check if `status.loadBalancer.ingress` exists right after a managed LB is created? If it does even though the IP address is still missing, then we could use that as a trigger.

The other possibility is to check the LB ID annotation, though using something on the status should be a bit more reliable.

Reply:
Can you answer this question of mine:

>Is req.OldObject possibly nil when the Service is brand new?

Reply:
I confirmed that oldObject is empty on a new service and added a condition to ensure that it is on create requests